File: microwave_topic/export_per_file_alarm_xlsx_test.py
import json
import tempfile
import unittest
from pathlib import Path

from openpyxl import load_workbook

from microwave_topic.export_per_file_alarm_xlsx import (
    export_jsonl_file,
    export_per_file_directory,
)


class ExportPerFileAlarmXlsxTest(unittest.TestCase):
    def setUp(self):
        self.group = {
            "uuid": "GROUP-1",
            "alarms": [
                {"告警编码ID": "A-2", "告警标题": "链路中断", "告警源": "NE-2"},
                {"告警编码ID": "A-1", "告警标题": "网元断连", "告警源": "NE-1"},
                {"告警编码ID": "A-4", "告警标题": "设备离线", "告警源": "NE-1"},
                {"告警编码ID": "A-3", "告警标题": "电源告警", "告警源": "NE-3"},
            ],
            "symptoms": [{"eid": "A-1", "alarm": "网元断连", "alarm_source": "NE-1"}],
            "ne_info": {
                "NE-1": {"site_id": "SITE-1", "manufacturer": "HUAWEI"},
                "NE-2": {"site_id": "SITE-1", "manufacturer": "HUAWEI"},
                "NE-3": {"site_id": "SITE-2", "manufacturer": "ZTE"},
            },
        }

    def test_exports_sorted_rows_and_merges_each_hierarchy_level(self):
        with tempfile.TemporaryDirectory() as tmpdir:
            input_path = Path(tmpdir) / "GROUP-1.jsonl"
            output_path = Path(tmpdir) / "GROUP-1.xlsx"
            input_path.write_text(json.dumps(self.group, ensure_ascii=False) + "\n", encoding="utf-8")
            self.assertEqual(export_jsonl_file(input_path, output_path), 4)
            workbook = load_workbook(output_path)
            sheet = workbook["告警明细"]

        self.assertEqual(
            [sheet.cell(1, column).value for column in range(1, 8)],
            [
                "站点ID",
                "设备厂家",
                "告警源",
                "告警标题",
                "告警首次发生时间",
                "告警清除时间",
                "告警最后发生时间",
            ],
        )
        self.assertEqual(sheet["A2"].value, "SITE-1")
        self.assertEqual(sheet["C2"].value, "NE-1")
        self.assertEqual(sheet["C4"].value, "NE-2")
        self.assertEqual(sheet["A5"].value, "SITE-2")
        self.assertIn("A2:A4", {str(item) for item in sheet.merged_cells.ranges})
        self.assertIn("B2:B4", {str(item) for item in sheet.merged_cells.ranges})
        self.assertIn("C2:C3", {str(item) for item in sheet.merged_cells.ranges})

    def test_writes_one_same_name_xlsx_per_jsonl(self):
        with tempfile.TemporaryDirectory() as tmpdir:
            input_dir = Path(tmpdir) / "jsonl"
            output_dir = Path(tmpdir) / "xlsx"
            input_dir.mkdir()
            (input_dir / "GROUP-1.jsonl").write_text(
                json.dumps(self.group, ensure_ascii=False) + "\n", encoding="utf-8"
            )
            stats = export_per_file_directory(input_dir, output_dir)
            self.assertTrue((output_dir / "GROUP-1.xlsx").exists())

        self.assertEqual(stats["input_file_count"], 1)
        self.assertEqual(stats["output_file_count"], 1)
        self.assertEqual(stats["alarm_count"], 4)


if __name__ == "__main__":
    unittest.main()

File: microwave_topic/export_per_file_alarm_xlsx_test.mjs
import assert from "node:assert/strict";
import fs from "node:fs/promises";
import os from "node:os";
import path from "node:path";
import test from "node:test";

import {
  exportJsonlFile,
  exportPerFileDirectory,
  loadArtifactTool,
} from "./export_per_file_alarm_xlsx.mjs";


const sampleGroup = {
  uuid: "GROUP-1",
  alarms: [
    {
      告警编码ID: "A-2",
      告警标题: "链路中断",
      告警源: "NE-2",
      告警首次发生时间: "2026-07-01 08:03:00",
    },
    {
      告警编码ID: "A-1",
      告警标题: "网元断连",
      告警源: "NE-1",
      告警首次发生时间: "2026-07-01 08:00:00",
      告警清除时间: "2026-07-01 08:10:00",
      扩展信息: { port: "1/1" },
    },
    {
      告警编码ID: "A-4",
      告警标题: "设备离线",
      告警源: "NE-1",
      告警首次发生时间: "2026-07-01 08:01:00",
      告警最后发生时间: "2026-07-01 08:09:00",
    },
    { 告警编码ID: "A-3", 告警标题: "电源告警", 告警源: "NE-3" },
  ],
  symptoms: [{ eid: "A-1", alarm: "网元断连", alarm_source: "NE-1" }],
  ne_info: {
    "NE-1": { site_id: "SITE-1", manufacturer: "HUAWEI", site_name: "站点一" },
    "NE-2": { site_id: "SITE-1", manufacturer: "HUAWEI", site_name: "站点一" },
    "NE-3": { site_id: "SITE-2", manufacturer: "ZTE", site_name: "站点二" },
  },
};


test("exports sorted hierarchical alarm workbook without duplicate visual alarms", async () => {
  const tempDir = await fs.mkdtemp(path.join(os.tmpdir(), "alarm-xlsx-"));
  const inputPath = path.join(tempDir, "GROUP-1.jsonl");
  const outputPath = path.join(tempDir, "GROUP-1.xlsx");
  await fs.writeFile(inputPath, `${JSON.stringify(sampleGroup)}\n`, "utf8");
  assert.equal(await exportJsonlFile(inputPath, outputPath), 4);

  const { FileBlob, SpreadsheetFile } = await loadArtifactTool();
  const workbook = await SpreadsheetFile.importXlsx(await FileBlob.load(outputPath));
  const inspected = await workbook.inspect({
    kind: "table",
    range: "告警明细!A1:H5",
    include: "values,formulas",
    tableMaxRows: 5,
    tableMaxCols: 8,
  });
  const output = inspected.ndjson;
  assert.match(output, /站点ID/);
  assert.match(output, /设备厂家/);
  assert.match(output, /告警源/);
  assert.match(output, /SITE-1/);
  assert.match(output, /NE-1/);
  assert.match(output, /NE-2/);
  assert.match(output, /SITE-2/);
});


test("writes one same-name xlsx for every per-file jsonl", async () => {
  const tempDir = await fs.mkdtemp(path.join(os.tmpdir(), "alarm-xlsx-dir-"));
  const inputDir = path.join(tempDir, "jsonl");
  const outputDir = path.join(tempDir, "xlsx");
  await fs.mkdir(inputDir);
  await fs.writeFile(path.join(inputDir, "GROUP-1.jsonl"), `${JSON.stringify(sampleGroup)}\n`, "utf8");
  const stats = await exportPerFileDirectory(inputDir, outputDir);
  assert.equal(stats.input_file_count, 1);
  assert.equal(stats.output_file_count, 1);
  assert.equal(stats.alarm_count, 4);
  await fs.access(path.join(outputDir, "GROUP-1.xlsx"));
});
